### ISSUE TITLE
Proposal to investigate Qubes as a replacement to Tails for the SVS

### ISSUE THREAD

**Issue by fowlslegs** [freedomofpress/securedrop#1151]
# Introduction

To begin, I'll admit I am just familiarizing myself with the SecureDrop system. I certainly have a lot to learn still in terms of the small-scale details of the system and a more full analysis (and certainly implementation) of my proposal would require much more familiarity with these details. I'd love to hear back from the devs who have been working on the project long-term and have this knowledge.
# Motivation

I think that the most recent [iSec audit](https://freedom.press/securedrop-files/iSEC_OTF_FPF_SecureDrop_Deliverable_v1.1_2.pdf) has brought up good points in appendices A.2 and A.3 (p. 25-6) regarding some weaknesses of Tails, especially in it’s capacity as the operating system running on the SVS. The authors write (emphasis mine):

> ...SecureDrop, due to its relatively simple functionality, and minimal processing of actual submissions, exposes a very limited attack surface to an attacker. However, this is less true of the [SVS] that uses Tails, where most of the processing of files submitted by anonymous sources will occur. Because the media uploaded by a source can take many forms (e.g., images, PDFs, videos, and office documents), _this leaves Tails with a wide attack surface exposed_.
> 
> [...]
> 
> The main risk that lies in the [SVS] is malware escaping the airgap via transfer media. Currently, since the [SVS] is typically set up with a persistent volume, this seems like a realistic threat, as malware can persist in the permanent volume, and spread via media used to transport documents. To avoid malware persisting in the secure view station, consider mounting the permanent volume read-only after its inital setup.

The issue of SVS persistence has also been discussed in issue #827. 

If it were me, I would write the last quoted paragraph slightly differently.  “The main risk” here lies not just in malware escaping the airgap, but in malware ever having the opportunity to modify any data that leaves the airgap.  Malware on the SVS need not infect the journalist workstation to compromise the anonymity of a source. Consider the following example:

Documents from a source are moved from the transfer device to the SVS, which has been persistently compromised. The malware collects various data and metadata about messages and files sent by the source before they are deleted or stripped with MAT. Then, when the journalist finishes their article and moves it onto the transfer device, the malware can embed the data it has collected via steganography into this departing document. The journalist publishes this article and our source is potentially compromised.

The solutions proposed in #827, including my own, as well as those in the audit are simply not sufficient and should only be considered as short-term solutions if SecureDrop truly seeks to defend against nation-state level adversaries. Over the last couple of years it has become abundantly clear that such adversaries can and do carry out advanced attacks such as persistent compromises via system BIOS, Option ROM, PCI EEPROM, or USB devices (the transfer device and possibly a printer, mouse, and keyboard). Privilege escalation in Linux, especially without GrSecurity, does not provide a very large obstacle to such adversaries.

One thing that is clear is that it nearly impossible to prevent malware on the SVS via malformed documents and media files. How then do we prevent this malware from ever leaving the SVS or from ever modifying any data that leaves the SVS?  As suggested in appendice A.2 of the audit, I believe the solution lies in isolation, or security by compartmentalization.

If Tails both implements application containers, [which it is exploring](https://tails.boum.org/blueprint/Linux_containers/#index1h2), and [MAC](https://tails.boum.org/blueprint/Mandatory_Access_Control/), this would be a huge improvement. At this point, however, this is a hypothetical; while we should consider the progress of the Tails hardening process as it continues, there are better solutions today. Considering all LiveUSB operating systems suitable of the SVS, I think Qubes OS stands out very strongly as a best candidate. In the rest of this issue, I will lay out why I think a transition of the SVS operating system from Tails to Qubes is one of the best ways, long-term, that SecureDrop can continue to increase the protection it offers to journalists and whistleblowers.
# Discussion

Qubes OS could best be described as a Xen hypervisor distribution. Qubes itself is the “glue” of the operating system; it is a clever integration process over isolated components. Qubes incorporates elements of security by compartmentalization, defense-in-depth, trusted computing, and by maintaining a well-audited, minimal TCB.  [Recently](https://groups.google.com/forum/#!msg/qubes-users/IQdCEpkooto/iyMh3LuzCAAJ), Qubes has released a LiveUSB. I don’t wish to explain the very basics of the operating system, when the [getting started documentation](https://www.qubes-os.org/en/doc/getting-started/) already does this pretty well.

There are a number of unique features Qubes offers, that are a boon not only to the security of the SVS, but to the Journalist Workstation as well. I think that a transition to Qubes would even allow a rewrite of several sections of the SecureDrop threat model that make various assumptions about uncompromised devices.

A transition to Qubes would not be without downsides, however, so those will be discussed as well. The biggest among them is that to get the most out of Qubes, one needs an 64-bit laptop with a TPM that supports Intel TXT and VT-D. I will point out when these technologies are needed when discussing the features of Qubes below.

A simple “pros vs. cons” list seems like a good way to illustrate the differences between the two distributions. Note that when you see “Qubes” below it is intended to mean “Qubes Live” with no additional software installed.
## Benefits of Qubes over Tails

• **Compartmentalization of data:** Qubes isolates VM data via a very minimal Xen image. QEMU, which many if of the most recent XSAs are related to, operates in a minimal, stub domain outside of dom0—thus placing it outside of the TCB.  Xen, especially without QEMU, has a better security track record than KVM and certainly LXC or Docker. Qubes also has a very minimal Dom0 image. If a journalist works on their article in a different VM than they open any files, there is a very, very low chance that Xen will be exploited and that their article may be infected with malware or used to exfiltrate sensitive information. Even if an adversary has root access in a VM, they will also be unable to attempt to reflash the BIOS or communicate with any of the hardware.  In fact, they would not even be able to identify any of the hardware on the machine.
• **Sanitize compromised documents:** Qubes has a tool, qvm-convert-pdf, which allows one to safely convert a potentially compromised PDF into a uncompromised PDF. It is explained in more detail [here](http://blog.invisiblethings.org/2013/02/21/converting-untrusted-pdfs-into-trusted.html), but, briefly, the PDF is opened in a disposable VM, where it is converted to RGB format, and then sent back to the orignal VM for reconstruction. Reconstruction is carried out by ImageMagick, and one may reasonably assume it is very hard to exploit the ImageMagick RGB parser. Movement of files is done over a minimal, extensible inter-VM communcation framework called [Qrexec](https://www.qubes-os.org/en/doc/qrexec3/), built over Xen shared memory. It should be noted, this is a one-click procedure via a pre-defined Nautlius action. This tool would allow journalists to work with PDFs (and anything that can be converted to a PDF including most document and image formats) in the same VM they write their articles. This allows the journalist to safely include images or pieces of documents shared by the source in their article and mostly eliminates the need to trust that the source’s computer is not compromised and that the source is not in fact an adversary.
• **Defense against malicious USBs:** On a computer with Intel VT-D (AMD-Vi is supported as well, I believe) it is possible to assign the USB controllers to a “USBVM.” This whole process can be carried out in about 5 clicks via the Qubes Manager GUI. This way, if the USB has malicious firmware, or if its partition tables are malformed to exploit some hypothetical overflow in the Linux kernel parsing stack, that compromise is compartmentalized to the USBVM. In this model, the transfer device should be encrypted, so that if the USBVM is compromised, it cannot see or modify the files on the USB, which will never be decrypted inside the USBVM. Though encrypting the USB device is not yet in the docs, @harlo is already training people to do so, and there is an issue about this, #1119.  Through the Qubes VM GUI one can then carry assign the USB device via Xen PCI passthrough to a VM where it can be safely decrypted. This can be done in two clicks and the device will be automounted in Nautilus, much like Tails, in the VM designated for decryption and storage of the files from the source.
• **Backup GPG and KeePassX keyrings w/ ease** A reliable, easy-to-use VM data export tool is available via the Qubes Manager GUI. In regards to the issues surrounding the ‘amnesia’ of LiveUSBs as discussed in #827, the first option @garrettr proposes could be carried out with just as much if not more ease in Qubes versus Tails. This would actually not require journalists to remember any additional passwords. Since Qubes is not persistent, that would be one less password to remember over the current model, and since you’d then have to remember the encrypted keyring storage USB password, that’d be one more—breaking even. Additionally, Qubes developers are interested enabling Tails-like persistent volumes in future versions (see original link about LiveUSB above).  Xen could then store the KeePassX and GPG keyrings in a persistent vault VM.
• **GPG keyring protection:** [Split-GPG](https://www.qubes-os.org/en/doc/split-gpg/) may currently be too difficult to set up and use for most non-technical users, but there is reportedly work on simplifying and integrating it into the OS. It is a clever tool aimed at protecting the compromise of your GPG key—see the link above for an explanation. It’s functionality could in some ways be mimicked, however, to prevent the private key from being compromised. After the keyring VM have been imported into Qubes, one would import the documents from the transfer device to the vault VM, preferably via a USBVM. The documents should be decrypted, but nothing more; the only processing that should take place in this VM is decryption. Next, the documents should be moved to a new VM via the qvm-move-to-vm tool, which is available via a Nautilus action. If a file were to be opened in the vault VM, it could compromise the VM and later exfiltrate the private key, _even if only sanitized documents ever leave the airgap_. This is because qubes-pdf-convert is only intended to remove potential malware from documents. Malware could embed the private key in an image that gets sanitized and later appears in an article.
   ◦ I’ll take a quick moment to diverge on this idea. No unsanitized document should be opened in a VM where there are documents from different sources or from the same source at a different time unless you never plan on letting any of the material in that VM leave that VM. One cannot trust that a sources codename has not been compromised and that they are now not an adversary attempting to figure out what the source leaked.
• **KeePassX keyring protection:** In Tails, the transfer device encryption password may be stolen from the clipboard. X offers [next to no](http://blog.invisiblethings.org/2011/04/23/linux-security-circus-on-gui-isolation.html) input device or clipboard isolation. Wayland might be better, but that’s probably 4 years away from hitting Tails. In our model, unless the vault VM is compromised (unlikely as only decryption of the transfer device as a virtual block filesystem via Xen PCI passthrough and decryption of the source .zip should occur there), the decryption password will be safe. One _could_ separate the KeePassX keyring from the GPG keyring by creating a vault and GPG VM to store them each in, respectively. If executed correctly, the security benefit from doing so is exceedingly low and would simply create more work and more room for error for the journalist. In fact, in the model described above KeePassX is not providing much of a function for us at all. It would be an insignificant loss of security to simply store the passwords in plaintext. On my personal machine I use KeePassX in my vault VM, however, I use a simple 4-digit pin instead of a strong password. I do this because (a) KeePassX is a nice organizational tool and (b) if I were to export my keyring out of my vault VM, I could do so safely by generating a strong passphrase first. For similar reasons, it may make sense to continue to use KeePassX to store the GPG and transfer device passwords, but to use a keyfile stored in the vault VM to unlock it.
• **Trusted computing:** I won’t going into much detail here, because these things would require persistence, but Qubes can verify the integrity of the dom0 and Xen images at startup using TXT. TXT even protects the bootup process from DMA attacks that might occur between the measurement of code by the TPM and its execution via the SENTER instruction, which relies on VT-D. It can also measure the boot process and store that measurement in the TPM to defend against BIOS, Option ROM, PCI EEPROM, and GRUB tampering. While persistence may be some time off, the LiveUSB will support Secure Boot soon, which, considering the fact we’re trusting the BIOS (don’t really have a choice unless we move the whole platform to Libreboot), is certainly a good thing.
## Benefits of Tails over Qubes

• **User friendliness and complexity:** Tails is more user friendly and less complex than Qubes. Most people have no idea what a hypervisor is (let alone a kernel). On the other hand, Qubes GUI integration is very well done. Once people get over the initial concept of working with several isolated OSs simultaneously, I believe it eat model that make various assumptions about uncompromised devicerequires little technical expertise to operate. The title bars of each window are color-coded and include the name of the VM the window belongs to. This makes it much harder to make isolation-compromising mistakes. All necessary system operations can be done with an intuitive GUI, Qubes Manager, and use of the command line should not be necessary. Correctly executing the protocols mentioned above doesn’t necessarily have to be so hard, and only four VMs (USB, vault, untrusted, trusted) are really necessary. To illustrate a work I’ve outlined a sample workflow:

(1) Boot Qubes
(2) Create a USB VM
(3) Physically insert the keyring storage device
(4) Decrypt the keyring storage device in the USBVM
(5) Import the vault VM
(6) Physically remove the keyring storage device
(7) Unlock the KeePassX keyring with the key stored in the vault VM
(8) Physically insert the transfer device
(9) Virtually attach the transfer device to the vault VM
(10) Decrypt the transfer device
(11) Copy the encrypted archive into the vault VM
(12) “Wipe” the encrypted archive from the transfer device
(13) Virtually detach the transfer device from the vault VM
(14) Decrypt the archive in the vault VM
(15) Create an untrusted VM
(16) Move the unencrypted archive to the untrusted VM
(17) Unzip the archive in the untrusted VM
(18) Convert any documents or images you would like to use in your article to PDFs
(19) Create a trusted VM
(20) Move all these PDFs you wish to use material from from the untrusted VM to the trusted VM
(21) Convert all these untrusted PDFs to trusted PDFs in the trusted VM
(22) Write your article in the trusted VM
(23) Run MAT on all these trusted PDFs
(24) Virtually attach the transfer device to the trusted VM
(25) Move the article(s) onto the VM
(26) Virtually detach the transfer device from the trusted VM
(27) Physically remove the transfer device
(28) Shut down Qubes

Steps 3-7 could be replaced with simply entering a decryption password once Qubes Live supports persistence. (2) should still be executed each time because it minimizes the risk of cross-contamination of malware between USB devices that might be attached to the SVS. Obviously, Tails is easier to use? How much easier? How likely are people to make mistakes? We’ll need to do some testing to really figure this out, but I believe people should be capable of understanding and accurately executing a workflow like this.
• **MAT is not installed by default:** Qubes Live comes with a Fedora 21 live image that has pretty much every program a journalist might need to work with documents and media except for MAT. This could probably change if we asked the Qubes developers nicely.
• **Tails has lesser hardware requirements:** Qubes is an 64-bit only OS and requires a minimum of 3-4G of RAM. These are not huge hurdles, however, if you want to take most advantage of the protections it offers you will also need a computer with VT-D (and in the future TXT and a TPM might be useful for the LiveUSB as well). VT-D is a much more rare feature, usually shipped with business laptops like Thinkpads and Latitudes. Without VT-D USBs are processed by Dom0 and if they have compromised malware or malformed partition tables, they have effectively compromised the entire system. The protection it would offer against malicious documents, however, still makes it worth using.
• **Qubes doesn’t have persistence:** I think this will change soon, but especially with VT-D, I don’t think a persistent vault VM would expose a significant attack surface for persistent malware.
• **Qubes doesn’t wipe RAM on shutdown:** One of the biggest advantages about Tails is that it wipes memory on shutdown by kexec’ing MemTest86+. Qubes doesn’t do this and that leaves it open to cold boot attacks. It may be possible to add this feature, but a port would be tricky because of the vastly different archictures of the two OSs. This is probably the biggest advantage for Tails.
• **Tails receives more attention:** Qubes is becoming increasingly popular, but a lot more people still use (and probably look at Tails). This means security holes are more likely to be found and you might argue that Tails has been more thoroughly audited.
# Conclusion

As I see it, here are the major problems—ranked in increasing order of severity—in the way of a transition to Qubes, which I do believe is the overall best way to rectify security holes in the current model of using a persistent Tails volume (or an amnesiac one, for that matter):

(1) No MAT (could be simply added)
(2) No persistence (could be fairly simply added, I believe)
(3) Retrain journalists (would be more of a task to retrain journalists than to
train new SecureDrop users witheat model that make various assumptions about uncompromised device Qubes)
(4) No memory wipe on shut down (I have a feeling this will be difficult to implement with
Qubes)

I propose we explore solutions to these problems. (2) is in the works, while (1) could be added at our request. So then we could experiment with (3) by creating a sample training document and then doing some field testing with some non-technologists. (4) Might be the hardest to implement, but we should certainly investigate it. A simpler solution may be to instruct journalists to keep a hammer handy and to know where the RAM is located on their computer. In case of emergency, they might simply neutralize the threat of cold boot attacks by repeatedly smashing the hammer into the relevant part of the computer.

This is certainly something to be looked into in the long-term. I appreciate any feedback and thank you for reading this proposal. I did not mean for it to get this long!

**Comment by garrettr** [comment]
GitHub tip: prepending a `#` before an issue # will make it auto-convert into a clickable link to that issue (e.g. #1119).

**Comment by garrettr** [comment]
This is a fantastic outline of the many potential security and usability benefits that we could achieve for the SVS by using Qubes. Thanks for writing this up!

We have been discussing potentially using Qubes for the SVS internally, and are currently adopting a "wait and see" approach. In our opinion, the most difficult thing about using Qubes is setting up the various "trust zones" and manually enforcing workflows based around these trust zones to achieve security through compartmentalization. We see two potential avenues towards improving this situation:

1. The upcoming release of Qubes (3.1) will have a configuration manager than can be used to automatically set up different configurations of AppVM's in Qubes. We think that using this would be easier and more maintainable than forking and creating a custom "SecureDrop-Qubes". Very much looking forward to the release of this functionality!
2. While many of the actions that we'd like to use (e.g. decryption in the Split GPG VM, open documents in a disposable VM) are available through Nautilus actions, we'd probably want to automate and enforce the "secure" workflow with a custom "reading room" client. This is probably fairly straightforward, and my plan was to mess with the idea some more after Qubes 3.1 is released.

**Comment by fowlslegs** [comment]
> GitHub tip: prepending a # before an issue # will make it auto-convert into a clickable link to that issue (e.g. #1119).

Thanks, I just forgot in this case. Think I got them all now.

> we'd probably want to automate and enforce the "secure" workflow with a custom "reading room" client.

This is a great idea! I've actually spent a decent amount of time trying to come up with a high-level design for this. While I know Qubes internals well, I realized I'm going to really need to familiarize myself with (a) the SecureDrop source and (b) the current problems journalists have been facing in terms of their workflow (you mentioned a couple issues last time we spoke). So, for now, my notes are sitting in a text file. If you have any thoughts/ ideas on the topic feel free to share. We can definitely use this as a brainstorming thread.
